File: crates/ferrogate-guardrails/src/adapters/fixture.rs
// Token4AI Cloud Attribution
// Developed by the commercial cloud service company represented by https://token4ai.cloud.
// Author: jamesduan (X: https://x.com/JamesDuanL)
// Created: 2026-07-18
// description: Recorded request/response fixture transport for deterministic adapter tests (#201).

//! [`FixtureTransport`] replays RECORDED request/response JSON exchanges for
//! the native adapters, so conformance and evaluation runs are deterministic
//! and never touch the network. The recorded wire shapes live in
//! `tests/fixtures/*.json` and mirror real Presidio / LLM-Guard payloads;
//! matching is by exact JSON equality of the request body, so any drift
//! between the adapter's serializer and the recorded contract fails loudly.
//!
//! Compiled only under `cfg(test)` or the opt-in `conformance` feature.

use super::{DetectorTransport, TransportReply};
use crate::conformance::PROBE_SECRET;
use crate::{DetectorError, DetectorErrorKind};
use async_trait::async_trait;
use serde::Deserialize;
use serde_json::Value;

/// Recorded Presidio analyzer exchanges (see `tests/fixtures/`).
pub const PRESIDIO_FIXTURES: &str = include_str!("../../tests/fixtures/presidio_exchanges.json");
/// Recorded LLM-Guard API exchanges (see `tests/fixtures/`).
pub const LLM_GUARD_FIXTURES: &str = include_str!("../../tests/fixtures/llm_guard_exchanges.json");

#[derive(Debug, Clone, Deserialize)]
struct RecordedFile {
    #[allow(dead_code)]
    #[serde(default)]
    adapter: Option<String>,
    exchanges: Vec<RecordedExchange>,
}

#[derive(Debug, Clone, Deserialize)]
struct RecordedExchange {
    /// Human-readable label for triage; unused at runtime.
    #[allow(dead_code)]
    #[serde(default)]
    name: Option<String>,
    /// Free-form provenance note; unused at runtime.
    #[allow(dead_code)]
    #[serde(default)]
    note: Option<String>,
    request: Value,
    #[serde(default)]
    response: Option<RecordedReply>,
    /// `"hang"` makes the transport never reply, so the adapter's own
    /// deadline enforcement is genuinely exercised.
    #[serde(default)]
    outcome: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
struct RecordedReply {
    status: u16,
    #[serde(default)]
    body: Option<Value>,
    /// Raw (possibly non-JSON) body for malformed-response cases.
    #[serde(default)]
    body_raw: Option<String>,
}

/// A [`DetectorTransport`] that answers each request with the recorded reply
/// whose request JSON matches exactly. An unmatched request is a loud
/// `Internal` error so a drifted wire format can never silently pass.
#[derive(Debug)]
pub struct FixtureTransport {
    exchanges: Vec<RecordedExchange>,
}

impl FixtureTransport {
    /// Load recorded exchanges from fixture JSON. `${PROBE_SECRET}` in the
    /// fixture text is substituted with the runtime synthetic probe secret
    /// (the fixture files never carry a contiguous secret-shaped token).
    pub fn from_recorded(fixture_json: &str) -> Self {
        let substituted = fixture_json.replace("${PROBE_SECRET}", PROBE_SECRET);
        let file: RecordedFile =
            serde_json::from_str(&substituted).expect("recorded fixture file is valid JSON");
        Self {
            exchanges: file.exchanges,
        }
    }

    /// The recorded Presidio analyzer exchange set.
    pub fn presidio_recorded() -> Self {
        Self::from_recorded(PRESIDIO_FIXTURES)
    }

    /// The recorded LLM-Guard API exchange set.
    pub fn llm_guard_recorded() -> Self {
        Self::from_recorded(LLM_GUARD_FIXTURES)
    }
}

#[async_trait]
impl DetectorTransport for FixtureTransport {
    async fn post_json(&self, body: &[u8]) -> Result<TransportReply, DetectorError> {
        let request: Value = serde_json::from_slice(body).map_err(|_| {
            DetectorError::new(
                DetectorErrorKind::Internal,
                "fixture transport received a non-JSON request body",
            )
        })?;
        let Some(exchange) = self
            .exchanges
            .iter()
            .find(|exchange| exchange.request == request)
        else {
            return Err(DetectorError::new(
                DetectorErrorKind::Internal,
                "no recorded fixture exchange matches the adapter request (wire drift?)",
            ));
        };
        if exchange.outcome.as_deref() == Some("hang") {
            // Never resolves: the adapter's deadline wrap must fire.
            std::future::pending::<()>().await;
            unreachable!("pending future resolved");
        }
        let Some(reply) = &exchange.response else {
            return Err(DetectorError::new(
                DetectorErrorKind::Internal,
                "recorded fixture exchange has neither response nor outcome",
            ));
        };
        let body = match (&reply.body, &reply.body_raw) {
            (Some(value), _) => serde_json::to_vec(value).map_err(|_| {
                DetectorError::new(
                    DetectorErrorKind::Internal,
                    "recorded fixture body failed to serialize",
                )
            })?,
            (None, Some(raw)) => raw.clone().into_bytes(),
            (None, None) => Vec::new(),
        };
        Ok(TransportReply {
            status: reply.status,
            body,
        })
    }
}

File: packages/guardrails/src/adapters/fixture.ts
/**
 * Recorded request/response fixture transport — port of
 * `ferrogate-guardrails::adapters::fixture`.
 *
 * Replays recorded JSON exchanges for deterministic, network-free adapter tests.
 * Matching is by exact JSON equality of the request body, so any drift between
 * the adapter serializer and the recorded contract fails loudly. `"hang"`
 * exercises the adapter's own deadline enforcement.
 */

import { PROBE_SECRET } from "../conformance.js";
import { DetectorError } from "../contract.js";
import type { DetectorTransport, TransportReply } from "./transport.js";

interface RecordedReply {
  status: number;
  body?: unknown;
  body_raw?: string;
}

interface RecordedExchange {
  name?: string;
  note?: string;
  request: unknown;
  response?: RecordedReply;
  outcome?: string;
}

interface RecordedFile {
  adapter?: string;
  exchanges: RecordedExchange[];
}

export class FixtureTransport implements DetectorTransport {
  private exchanges: RecordedExchange[];

  private constructor(exchanges: RecordedExchange[]) {
    this.exchanges = exchanges;
  }

  /** Load from recorded JSON (string or parsed object). `${PROBE_SECRET}` is substituted. */
  static fromRecorded(fixture: string | RecordedFile): FixtureTransport {
    let file: RecordedFile;
    if (typeof fixture === "string") {
      file = JSON.parse(fixture.replaceAll("${PROBE_SECRET}", PROBE_SECRET)) as RecordedFile;
    } else {
      file = JSON.parse(
        JSON.stringify(fixture).replaceAll("${PROBE_SECRET}", PROBE_SECRET),
      ) as RecordedFile;
    }
    return new FixtureTransport(file.exchanges);
  }

  async postJson(body: Uint8Array): Promise<TransportReply> {
    let request: unknown;
    try {
      request = JSON.parse(new TextDecoder().decode(body));
    } catch {
      throw DetectorError.new("internal", "fixture transport received a non-JSON request body");
    }
    const exchange = this.exchanges.find((e) => deepEqual(e.request, request));
    if (!exchange) {
      throw DetectorError.new(
        "internal",
        "no recorded fixture exchange matches the adapter request (wire drift?)",
      );
    }
    if (exchange.outcome === "hang") {
      return new Promise<TransportReply>(() => {
        /* never resolves: the adapter's deadline wrap must fire */
      });
    }
    const reply = exchange.response;
    if (!reply) {
      throw DetectorError.new(
        "internal",
        "recorded fixture exchange has neither response nor outcome",
      );
    }
    let bodyBytes: Uint8Array;
    if (reply.body !== undefined) {
      bodyBytes = new TextEncoder().encode(JSON.stringify(reply.body));
    } else if (reply.body_raw !== undefined) {
      bodyBytes = new TextEncoder().encode(reply.body_raw);
    } else {
      bodyBytes = new Uint8Array();
    }
    return { status: reply.status, body: bodyBytes };
  }
}

function deepEqual(a: unknown, b: unknown): boolean {
  if (a === b) {
    return true;
  }
  if (typeof a !== typeof b || a === null || b === null) {
    return false;
  }
  if (Array.isArray(a) && Array.isArray(b)) {
    return a.length === b.length && a.every((item, i) => deepEqual(item, b[i]));
  }
  if (typeof a === "object" && typeof b === "object") {
    const ak = Object.keys(a as object);
    const bk = Object.keys(b as object);
    return (
      ak.length === bk.length &&
      ak.every((k) =>
        deepEqual((a as Record<string, unknown>)[k], (b as Record<string, unknown>)[k]),
      )
    );
  }
  return false;
}
